### DISCUSSION TITLE
how to assert with respect to traced values inside vmap?

### DISCUSSION THREAD

**Discussion by aldopareja** [jax-ml/jax#12785] [Q&A] [answered]
Sorry if this has been asked before. This is what I want, this way I can make sure the function is never used under unexpected conditions.

```python
def check_value(a):
    assert a.shape == (1,) #works
    assert a>0 #doesn't work

vmap(check_value)(jnp.ararge(10)-5)
```

**Comment by jakevdp** [comment]
You cannot write assertions with respect to traced values in JAX transforms (`jit`, `vmap`, etc.).

There's some more background on this in [Common Gotchas: Python Control Flow](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#python-control-flow-jit) – this section specifically discusses control flow, but assertions are basically control flow in disguise (`assert a>0` is basically `if a>0: raise AssertionError()`).

The [How To Think In JAX](https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html) document is also some good background reading to understand why JAX behaves this way.

There may be workarounds depending on what you were trying to do originally: can you say more about what your goal was here? `checkify` and `host_callback` are both possible approaches to achieve similar results, depending on the context and how important performance is for your function.

**Comment by aldopareja** [comment, marked as answer]
Given the debugging tools that have spawned recently (e.g `jax.debug.print`), is there a way we can assert with respect to the values of traced arrays?

**Reply by jakevdp** [reply]
Sure, for example, you could use `debug.callback` to do a host-side assertion of the value:
```python
import jax

def check_value(val):
  assert val > 0

def f(x):
  jax.debug.callback(check_value, x)
  return x

jax.vmap(f)(jax.numpy.array([1, 2, 3]))
# DeviceArray([1, 2, 3], dtype=int32)

jax.vmap(f)(jax.numpy.array([1, 2, -1]))
# AssertionError
```
This is probably not to be recommended in normal production code (the callback is done synchronously, so will likely make any real code much slower) but for debugging purposes it's a decent approach.

There's also [`jax.experimental.checkify`](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.checkify.checkify.html) which may eventually become a good approach for this kind of thing.

**Reply by aldopareja** [reply]
thank you so much!
